Rewrite encoding detection

scx opened the pull request:
Detect UTF-8 BOM when loading ass file.

This should make loading UTF-8-BOM files faster.
This also guarentee Aegisub will load large ass saved by itself.

See Aegisub#110

https://github.com/wangqr/Aegisub/commit/3502fd112c06d38ec17d6f9d5035ab56203c355d

---

Rewrite encoding detection.

Now feeds all data to uchardet, when uchardet is available. The file size limit is removed.

When uchardet is not available, we look for UTF-8 BOM.
This should make loading UTF-8-BOM files faster.
Because Aegisub always save file in UTF-8-BOM, this should also guarentee Aegisub will load large (>100MB) file saved by itself.

See Aegisub/Aegisub#110

https://github.com/wangqr/Aegisub/commit/f733297499f72ea11c166f6be23751a7b002c51c

FichteFoll commented on libaegisub/common/charset.cpp:
```diff
@@ -29,37 +29,40 @@ namespace agi { namespace charset {
 #else
-	auto read = std::min<uint64_t>(4096, fp.size());
+
+	// Look for utf-8 BOM
```
This block is duplicated because it's also before the uchardet block.

scx replied:
Fair point.

scx commented:
I just reverted the duplicate patch and added two more, originally created by @wangqr.
https://github.com/wangqr/Aegisub/commit/d6ddea0f6577df319b5f2cbd2a54d5206693f752
https://github.com/wangqr/Aegisub/commit/fbca22229524210c3c9ae0f2bb423d405065d8f0#diff-db8148e40f1f9cbc536fb61329821419

tgoyne commented:
Unconditionally feeding the entire file into uchardet seems like it'd be pointlessly slow for large files. Are the any examples of files where feeding it data until it thinks it knows what the encoding is and stopping there gives incorrect results?

wangqr commented:
The open file dialog already has a limit on file extensions. Among them only MKV is non-text file format and could be big. So it is separately tested.

I've timed loading a 1GB ass on my i7-8750H and SK hynix 256GB NVMe SSD, it takes 29s and is disk read bounded. So without this PR, only files less than 100MB can be loaded, and they will be loaded within 3s even after the PR.

wangqr commented:
As for incorrect result, https://github.com/Aegisub/Aegisub/issues/121#issuecomment-498407462 is an example IIRC.
